### ISSUE TITLE
Execute scripts preversion, version and postversion during lerna publish.

### ISSUE THREAD

**Issue by ncochard** [lerna/lerna#774]
Hello

First of all, thank you for `lerna`, awesome tool.

The command `npm version ...` will execute the npm scripts `preversion`, `version` and `postversion`. See this link for more details: [https://docs.npmjs.com/misc/scripts](https://docs.npmjs.com/misc/scripts).

> * preversion: Run BEFORE bumping the package version.
> * version: Run AFTER bumping the package version, but BEFORE commit.
> * postversion: Run AFTER bumping the package version, and AFTER commit.

It would be great if the command `lerna publish` could execute these scripts when increasing the version number of a package.

**Comment by noherczeg** [comment]
Lerna runs the `npm publish` command internally. I believe it should work currently as-is.

On the other hand `npm publish` does not run `version` internally if I'm not mistaken.

**Comment by ncochard** [comment]
That is correct: `npm publish` does not run `npm version` internally.

Let me try to clarify my suggestion...

The typical workflow in a pure *npm* project would be the following:

1. We fist execute `npm version` to increase the version number and create a git tag.
   (This also executes our custom npm scripts `npm run preversion`, `npm run version` and `npm run postversion`.)
2. Then we can execute `npm publish`.
   (This also execute `npm run prepublish`, `npm run prepublishOnly`, etc.)

Our workflow is simplified with lerna because the command `lerna publish` figuratively speaking *encapsulates* the `npm version` command and the `npm publish` command.

Since lerna already honors the script `prepublish`; for consistency, it would be useful if it also honored the scripts `version`. This would allow us to migrate npm projects to lerna and benefit from the same features.

I can understand however that it is probably not a high priority feature: the scripts `preversion`, `version` and `postversion` are not often used.

**Comment by evocateur** [comment]
I'm in favor of supporting lifecycle scripts whenever we can. I don't envision actually calling `npm version` because of all the side effects, but running the scripts at the appropriate times seems very doable.

**Comment by cwmoo740** [comment]
@evocateur 

These lifecycle scripts are currently non-blocking, so things might not work 100% as expected. For example, if a preversion script fails, it doesn't halt the commit, and if a preversion script adds files to be committed, they might not make it in time.

I can open another PR to make them wait on the scripts if that's worthwhile.
